physics/cu_unified_deep.F90

```diff
@@ -0,0 +1,6051 @@
+!$acc declare copy(entr_rate_2d,zuo,ierr,ktop) copyin(p_cup, heo,heso_cup,z_cup,hkbo,kstabi,k22,kbcon,kpbl,klcl)
+     real(kind=kind_phys), dimension (its:ite,kts:kte) :: hcot
+!$acc declare create(hcot)
+     character *(*), intent (in) ::    name
+     real(kind=kind_phys) :: dz,dh, dbythresh
+     real(kind=kind_phys) :: dby(kts:kte)
+     integer :: i,k,ipr,kdefi,kstart,kbegzu,kfinalzu
+     integer, dimension (its:ite) :: start_level
+!$acc declare create(start_level)
+     integer,parameter :: find_ktop_option = 1 !0=original, 1=new
+     
+     dbythresh=0.8 !0.95  ! the range of this parameter is 0-1, higher => lower
+                    ! overshoting (cheque aa0 calculation)
+                    ! rainfall is too sensible this parameter
+                    ! for now, keep =1.
+     if(name == 'shallow'.or. name == 'mid')then 
+         dbythresh=1.0
+     endif
+     !         print*,"================================cumulus=",name; call flush(6)
+!$acc parallel loop private(dby,kfinalzu,dz)
+     do i=its,itf
+       kfinalzu=ktf-2
+       ktop(i)=kfinalzu
+       if(ierr(i).eq.0)then
+       dby (kts:kte)=0.0
+
+       start_level(i)=kbcon(i)
+       !-- hcot below kbcon
+       hcot(i,kts:start_level(i))=hkbo(i)
+
+       dz=z_cup(i,start_level(i))-z_cup(i,start_level(i)-1)
+       dby(start_level(i))=(hcot(i,start_level(i))-heso_cup(i,start_level(i)))*dz
+       
+       !print*,'hco1=',start_level(i),kbcon(i),hcot(i,start_level(i))/heso_cup(i,start_level(i))
+!$acc loop seq
+       do k=start_level(i)+1,ktf-2
+           dz=z_cup(i,k)-z_cup(i,k-1)
+
+           hcot(i,k)=( (1.-0.5*entr_rate_2d(i,k-1)*dz)*hcot(i,k-1)    &
+                              +entr_rate_2d(i,k-1)*dz *heo (i,k-1) )/ &
+                       (1.+0.5*entr_rate_2d(i,k-1)*dz)
+           dby(k)=dby(k-1)+(hcot(i,k)-heso_cup(i,k))*dz
+           !print*,'hco2=',k,hcot(i,k)/heso_cup(i,k),dby(k),entr_rate_2d(i,k-1)
+
+       enddo
+       if(find_ktop_option==0) then 
+        do k=maxloc(dby(:),1),ktf-2
+          !~ print*,'hco30=',k,dby(k),dbythresh*maxval(dby)
+      
+          if(dby(k).lt.dbythresh*maxval(dby))then
+              kfinalzu = k - 1
+              ktop(i)  = kfinalzu
+              !print*,'hco4=',k,kfinalzu,ktop(i),kbcon(i)+1;call flush(6)
+              go to 412
+          endif
+        enddo
+        412    continue
+       else
+         do k=start_level(i)+1,ktf-2
+          !~ print*,'hco31=',k,dby(k),dbythresh*maxval(dby)
+      
+          if(hcot(i,k) < heso_cup(i,k) )then
+              kfinalzu = k - 1
+              ktop(i)  = kfinalzu
+              !print*,'hco40=',k,kfinalzu,ktop(i),kbcon(i)+1;call flush(6)
+              exit
+          endif
+        enddo
+       endif
+       if(kfinalzu.le.kbcon(i)+1) ierr(i)=41
+       !~ print*,'hco5=',k,kfinalzu,ktop(i),kbcon(i)+1,ierr(i);call flush(6)
+      !
+      endif
+     enddo
+!$acc end parallel
+  end subroutine get_cloud_top
+
```

Comment:
@dustinswales @grantfirl @Qingfu-Liu the subroutine calculate updraft velocity is new

Reply:
The newly added code looks good to me. The old code "function satvap(temp2)" in cu_gf_deep.F90 can be made more efficient since it is used for all model grid points. The code may be changed in the future.

Reply:
@Qingfu-Liu thank you for this suggestion. I will make a note!